### PR TITLE
[JENKINS-72627] Improve locale parsing for loading of localised help files

### PR DESCRIPTION
**Pull request by BobDu**
See [JENKINS-72627](https://issues.jenkins.io/browse/JENKINS-72627).

### Testing done

Add some automated unit testing.

Also, manually tested it.

1. Install [dingtalk plugin](https://plugins.jenkins.io/dingding-notifications/), version `2.7.1`
2. run cmd `curl -H "Accept-Language:zh,zh-CN;q=0.9,en;q=0.8" http://localhost:8080/jenkins/descriptor/io.jenkins.plugins.DingTalkRobotConfig/help/id`

Before modify, return 404 page
After modify, return `help-id_zh_CN.html` text.
```
机器人 id，可以在 pipeline 中使用；留空会自动生成，手动填写请保证唯一性
```

PS: Because the `help-id_zh_CN.html` file is missing in version `2.7.1` of `dingtalk-plugin` , It is reasonable to return a 404 before modify.

```
# in ~/dingtalk-plugin/src/main on git:main x [12:31:30] 
$ tree resources 
resources
├── index.jelly
└── io
    └── jenkins
        └── plugins
 ...
            ├── DingTalkRobotConfig
            │   ├── config.jelly
            │   ├── help-id_zh_CN.html
            │   ├── help-key_zh_CN.html
            │   └── help-name_zh_CN.html

```

### Proposed changelog entries

- JENKINS-72627, Improve locale parsing for loading of localised help files

```[tasklist]
### Submitter checklist
- [x] The Jira issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [x] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

File: core/src/main/java/hudson/model/Descriptor.java
Comment (NotMyFault) [inline core/src/main/java/hudson/model/Descriptor.java:60]: Please revert, we don't use asterisk imports.

**Review by timja — COMMENTED** [review]
Is it possible to write a test that verifies it works for this locale and others?

**Comment by BobDu** [comment]
> Is it possible to write a test that verifies it works for this locale and others?

Good idea. I believe it will be better with unit tests. I will try to add it.

**Review by timja — COMMENTED** [review]
thanks, here's some feedback

**Comment by Wiiscord-GH** [comment]
ly

File: core/src/test/java/hudson/GetLocaleStaticHelpUrlTest.java
Comment (timja) [inline core/src/test/java/hudson/GetLocaleStaticHelpUrlTest.java:217]: one nit I would add assert to the beginning of this method:
```suggestion
    private void assertThatLocaleResourceIs(URL url, String resource) {
```

**Review by timja — APPROVED** [review]
Thanks for the unit tests, much clearer and provides good coverage!

LGTM, one test naming nit but not a blocker

**Comment by timja** [comment]
Can you please restore the PR template in full, fill out testing done section and add a changelog entry

https://github.com/jenkinsci/jenkins/blob/master/.github/PULL_REQUEST_TEMPLATE.md

**Comment by BobDu** [comment]
> Can you please restore the PR template in full, fill out testing done section and add a changelog entry

@timja Done.

**Review by NotMyFault — APPROVED** [review]
Thanks!

**Comment by NotMyFault** [comment]
/label ready-for-merge

---

This PR is now ready for merge. We will merge it after ~24 hours if there is no negative feedback.
Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process.
Thanks!

**Comment by mawinter69** [comment]
I think this change causes some problems.
I have chrome and firefox configured to use English as my first language. But I also have German as additional language.
Now the UI is shown in English consistently (those texts defined via the property files) but whenever a help is available in german it is shown in german. Only when I remove German as language the help is shown in English. 
I would expect that when I choose English as main language that everything is shown in English

**Comment by mawinter69** [comment]
Additionally this change affects the locale plugin, as it only intercepts getLocale() and not getLocales()

**Comment by BobDu** [comment]
Can you open chrome dev tools can check your page request header `Accept-Language` ?

@mawinter69

**Comment by mawinter69** [comment]
`en-US,en;q=0.9,de;q=0.8`

It is clear why this happens. We do not have explicit english texts, so we will never match for `en`  but rely on the fallback with no language. As we now loop over all accepted languages, `en` doesn't match but `de` matches and we get german help choosen.
We will probably need special treatment if `en` is the preferred language

**Comment by BobDu** [comment]
> We will probably need special treatment if en is the preferred language
+1

**Comment by MarkEWaite** [comment]
> I think this change causes some problems.
> I have chrome and firefox configured to use English as my first language. But I also have German as additional language.
> Now the UI is shown in English consistently (those texts defined via the property files) but whenever a help is available in german it is shown in german. Only when I remove German as language the help is shown in English.

Thanks for explaining something that I had observed but did not understand well enough to report as an issue.  I have Italian configured as an additional language and was surprised that I was seeing Italian language help text in some cases.

**Comment by jonesbusy** [comment]
> Additionally this change affects the locale plugin, as it only intercepts getLocale() and not getLocales() @jonesbusy

Tentative fix on https://github.com/jenkinsci/locale-plugin/pull/222

To be honest I'm not sure to understand the expected behavior with several local so any help is appreciated

**Comment by BobDu** [comment]
> I have filed [JENKINS-73246](https://issues.jenkins.io/browse/JENKINS-73246) to track this regression and confirmed that the regression is solved by reverting this PR. In the absence of a fix by Friday, June 7, 2024, I will revert commit [abf3b41](https://github.com/jenkinsci/jenkins/commit/abf3b41ab33f34436cb61b84fe3d5f03d53a6bd5) to restore stability to the default branch.

I am working on preparing a further PR to fix this issue, but it might not be completed by June 7th. Could we possibly extend the deadline by 1-2 weeks?

**Comment by basil** [comment]
Sure, I will hold off on the revert for a bit longer.

**Pull request merged**
